### PR TITLE
s3: Support requests xml body with non utf8 encoding

### LINKED ISSUE #6898: Oracle RMAN hangs when using Minio for Azure
1. Created Minio application service using the steps described here https://open.microsoft.com/2017/11/09/s3cmd-amazon-s3-compatible-apps-azure-storage/

```
$ az appservice plan create \
    --name "MinioAppPlan" \
    --is-linux \
    --sku B1 \
    --resource-group "mari184V2" \
    --location "CentralUS"
```

```
$ az webapp create \
    --name "aleminio" \
    --deployment-container-image-name "minio/minio" \
    --plan "MinioAppPlan" \
    --resource-group "mari184V2"
```
	
```
$ az webapp config appsettings set \
    --settings "MINIO_ACCESS_KEY=..." "MINIO_SECRET_KEY=..." "PORT=9000" \
    --name "aleminio" \
    --resource-group "mari184V2"
```

```
$ az webapp config set \
    --startup-file "gateway azure" \
    --name "aleminio" \
    --resource-group "mari184V2"
```
	
2. Installed Oracle Secure Backup module using the following steps: https://blog.pythian.com/backup-oracle-databases-aws-s3/

`$ /u01/app/oracle/product/18.3.0/dbhome_1/jdk/bin/java -jar osbws_install.jar -AWSID ... -AWSKey ... -otnUser mariam.kupa@gmail.com     -walletDir $ORACLE_HOME/dbs/osbws_wallet     -libDir $ORACLE_HOME/lib -awsEndPoint aleminio.azurewebsites.net  -location maribucket
`
The same module works with AWS and I am able to back up the database. But it hangs when using Minio.

RMAN script that is working with AWS but hangs with Minio is the following:

```
run{
  allocate channel c1 device type sbt parms='SBT_LIBRARY=libosbws.so,SBT_PARMS=(OSB_WS_PFILE=/u01/app/oracle/product/18.3.0/dbhome_1/dbs/osbwsorcl1.ora)';
  backup current controlfile;
  }
```
  
Where /u01/app/oracle/product/18.3.0/dbhome_1/dbs/osbwsorcl1.ora file contains:

```
OSB_WS_HOST=http://aleminio.azurewebsites.net
#OSB_WS_LOCATION=maribucket
OSB_WS_BUCKET=maribucket
OSB_WS_VIRTUAL_HOST=FALSE
OSB_WS_WALLET='location=file:/u01/app/oracle/product/18.3.0/dbhome_1/dbs/osbws_wallet CREDENTIAL_ALIAS=_aws'
```

After some time the following error is returned by RMAN:

```
ORA-27192: skgfcls: sbtclose2 returned error - failed to close file
ORA-19511: non RMAN, but media manager or vendor specific failure, error text:
   LHT-00027: Key not found in hash table.
LHT-00027: Key not found in hash table.
LHT-00027: Key not found in hash table.
```

During RMAN backup I have noticed that metadata files are created in the Azure container through Minio, moreover from database I see that RMAN reads datafile but is not able to write backup into Azure container. 

How can I trace what happens on Minio application service, why it hangs?

### PR DESCRIPTION
**Pull request by vadmeste**
## Description
One user reported having discovered the following error:

API: SYSTEM()
Time: 20:06:17 UTC 12/06/2018
Error: xml: encoding "US-ASCII" declared but Decoder.CharsetReader is nil
1: cmd/handler-utils.go:43:cmd.parseLocationConstraint()
2: cmd/auth-handler.go:250:cmd.checkRequestAuthType()
3: cmd/bucket-handlers.go:411:cmd.objectAPIHandlers.PutBucketHandler()
4: cmd/api-router.go100cmd.(objectAPIHandlers).PutBucketHandler-fm()
5: net/http/server.go:1947:http.HandlerFunc.ServeHTTP()

Hence, adding support of different xml encoding. Although there
is no clear specification about it, manual tests has shown that
setting US-ASCII in xml body in Put Bucket request doesn't change
the normal behavior of Put Bucket API.

## Motivation and Context
Fixes #6898 

## Regression
No

## How Has This Been Tested?
Difficult to test (Oracle RMAN is painful to install in ubuntu/debian distro)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
```
Error running mint automation
+ mkdir -p 6953-ce61296/gopath/src/github.com/minio
+ git clone --quiet https://github.com/minio/minio.git 6953-ce61296/gopath/src/github.com/minio/minio
++ cd 6953-ce61296/gopath/src/github.com/minio/minio
++ git remote add minio https://github.com/minio/minio.git
++ cd 6953-ce61296/gopath/src/github.com/minio/minio
++ git fetch --quiet minio pull/6953/head:pr6953
++ cd 6953-ce61296/gopath/src/github.com/minio/minio
++ git checkout --quiet pr6953
++ minikube ip
+ minikube_ip=192.168.39.96
++ mkdir -p certs
++ cd certs
++ ./cert -host 192.168.39.96
2018/12/11 16:14:20 wrote public.crt
2018/12/11 16:14:20 wrote private.key
+ GOPATH=/mnt/nvme0n1p1/mint-auto/mint-auto/6953-ce61296/gopath
+ make -C 6953-ce61296/gopath/src/github.com/minio/minio --quiet
cmd/utils.go:42:2: cannot find package "golang.org/x/net/html/charset" in any of:
	/mnt/nvme0n1p1/mint-auto/mint-auto/6953-ce61296/gopath/src/github.com/minio/minio/vendor/golang.org/x/net/html/charset (vendor tree)
	/mnt/nvme0n1p1/mint-auto/.gimme/versions/go1.10.4.linux.amd64/src/golang.org/x/net/html/charset (from $GOROOT)
	/mnt/nvme0n1p1/mint-auto/mint-auto/6953-ce61296/gopath/src/golang.org/x/net/html/charset (from $GOPATH)
make: *** [build] Error 1

```

**Comment by minio-ops** [comment]
```
Error running mint automation
+ mkdir -p 6953-9696b16/gopath/src/github.com/minio
+ git clone --quiet https://github.com/minio/minio.git 6953-9696b16/gopath/src/github.com/minio/minio
++ cd 6953-9696b16/gopath/src/github.com/minio/minio
++ git remote add minio https://github.com/minio/minio.git
++ cd 6953-9696b16/gopath/src/github.com/minio/minio
++ git fetch --quiet minio pull/6953/head:pr6953
++ cd 6953-9696b16/gopath/src/github.com/minio/minio
++ git checkout --quiet pr6953
++ minikube ip
+ minikube_ip=192.168.39.96
++ mkdir -p certs
++ cd certs
++ ./cert -host 192.168.39.96
2018/12/11 16:15:09 wrote public.crt
2018/12/11 16:15:09 wrote private.key
+ GOPATH=/mnt/nvme0n1p1/mint-auto/mint-auto/6953-9696b16/gopath
+ make -C 6953-9696b16/gopath/src/github.com/minio/minio --quiet
vendor/golang.org/x/net/html/charset/charset.go:19:2: cannot find package "golang.org/x/net/html" in any of:
	/mnt/nvme0n1p1/mint-auto/mint-auto/6953-9696b16/gopath/src/github.com/minio/minio/vendor/golang.org/x/net/html (vendor tree)
	/mnt/nvme0n1p1/mint-auto/.gimme/versions/go1.10.4.linux.amd64/src/golang.org/x/net/html (from $GOROOT)
	/mnt/nvme0n1p1/mint-auto/mint-auto/6953-9696b16/gopath/src/golang.org/x/net/html (from $GOPATH)
vendor/golang.org/x/net/html/charset/charset.go:20:2: cannot find package "golang.org/x/text/encoding" in any of:
	/mnt/nvme0n1p1/mint-auto/mint-auto/6953-9696b16/gopath/src/github.com/minio/minio/vendor/golang.org/x/text/encoding (vendor tree)
	/mnt/nvme0n1p1/mint-auto/.gimme/versions/go1.10.4.linux.amd64/src/golang.org/x/text/encoding (from $GOROOT)
	/mnt/nvme0n1p1/mint-auto/mint-auto/6953-9696b16/gopath/src/golang.org/x/text/encoding (from $GOPATH)
vendor/golang.org/x/net/html/charset/charset.go:21:2: cannot find package "golang.org/x/text/encoding/charmap" in any of:
	/mnt/nvme0n1p1/mint-auto/mint-auto/6953-9696b16/gopath/src/github.com/minio/minio/vendor/golang.org/x/text/encoding/charmap (vendor tree)
	/mnt/nvme0n1p1/mint-auto/.gimme/versions/go1.10.4.linux.amd64/src/golang.org/x/text/encoding/charmap (from $GOROOT)
	/mnt/nvme0n1p1/mint-auto/mint-auto/6953-9696b16/gopath/src/golang.org/x/text/encoding/charmap (from $GOPATH)
vendor/golang.org/x/net/html/charset/charset.go:22:2: cannot find package "golang.org/x/text/encoding/htmlindex" in any of:
	/mnt/nvme0n1p1/mint-auto/mint-auto/6953-9696b16/gopath/src/github.com/minio/minio/vendor/golang.org/x/text/encoding/htmlindex (vendor tree)
	/mnt/nvme0n1p1/mint-auto/.gimme/versions/go1.10.4.linux.amd64/src/golang.org/x/text/encoding/htmlindex (from $GOROOT)
	/mnt/nvme0n1p1/mint-auto/mint-auto/6953-9696b16/gopath/src/golang.org/x/text/encoding/htmlindex (from $GOPATH)
make: *** [build] Error 1

```

**Comment by harshavardhana** [comment]
> One user reported having discovered the following error:
> 
> API: SYSTEM()
> Time: 20:06:17 UTC 12/06/2018
> Error: xml: encoding "US-ASCII" declared but Decoder.CharsetReader is nil
> 1: cmd/handler-utils.go:43:cmd.parseLocationConstraint()
> 2: cmd/auth-handler.go:250:cmd.checkRequestAuthType()
> 3: cmd/bucket-handlers.go:411:cmd.objectAPIHandlers.PutBucketHandler()
> 4: cmd/api-router.go100cmd.(objectAPIHandlers).PutBucketHandler-fm()
> 5: net/http/server.go:1947:http.HandlerFunc.ServeHTTP()
> 
> Hence, adding support of different xml encoding. Although there
> is no clear specification about it, manual tests has shown that
> setting US-ASCII in xml body in Put Bucket request doesn't change
> the normal behavior of Put Bucket API.

UTF-8 is expected for all XMLs in S3 API. Its a bug in Oracle RMAN we should close the issue and not support different encodings without proper specification from AWS S3.

**Comment by vadmeste** [comment]
> UTF-8 is expected for all XMLs in S3 API. Its a bug in Oracle RMAN we should close the issue and not support different encodings without proper specification from AWS S3.

The thing is that Oracle RMAN seems to work with AWS S3, also I tested Put Bucket with us-ascii encoding and AWS didn't complain

**Comment by harshavardhana** [comment]
> > UTF-8 is expected for all XMLs in S3 API. Its a bug in Oracle RMAN we should close the issue and not support different encodings without proper specification from AWS S3.
> 
> The thing is that Oracle RMAN seems to work with AWS S3, also I tested Put Bucket with us-ascii encoding and AWS didn't complain

did you try this with different regions?

**Comment by vadmeste** [comment]
> did you try this with different regions?
Yes,


```
PUT / HTTP/1.1
Host: vadmeste-us-east-5.s3-us-east-2.amazonaws.com
User-Agent: Minio (linux; amd64) minio-go/v6.0.12
Content-Length: 195
Authorization: AWS4-HMAC-SHA256 Credential=xxxx/20181211/us-east-2/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=xxxxxxxx
Content-Md5: YaEBXdb1zhBKZJogYlTNiA==
X-Amz-Content-Sha256: bb098152e47cc80a0f44281e8dbc081c195edfc52455b1500d435a057e80c4af
X-Amz-Date: 20181211T223409Z

<?xml version="1.0" encoding="US-ASCII"?>
<CreateBucketConfiguration xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><LocationConstraint>us-east-2</LocationConstraint></CreateBucketConfiguration>

HTTP/1.1 200 OK
x-amz-id-2: wFyFV2zNkBHpv2Zlldr+VROnnt4xTBP643l3SV7e9dq/l0oZgL97tq6PAapIO97HPC3g6zGjr80=
x-amz-request-id: 9071F6FE21D5B220
Date: Tue, 11 Dec 2018 22:34:10 GMT
Location: http://vadmeste-us-east-5.s3.amazonaws.com/
Content-Length: 0
Server: AmazonS3
```

**Comment by harshavardhana** [comment]
> PUT / HTTP/1.1 Host: vadmeste-us-east-5.s3-us-east-2.amazonaws.com User-Agent: Minio (linux; amd64) minio-go/v6.0.12 Content-Length: 195 Authorization: AWS4-HMAC-SHA256 Credential=xxxx/20181211/us-east-2/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=xxxxxxxx Content-Md5: YaEBXdb1zhBKZJogYlTNiA== X-Amz-Content-Sha256: bb098152e47cc80a0f44281e8dbc081c195edfc52455b1500d435a057e80c4af X-Amz-Date: 20181211T223409Z <?xml version="1.0" encoding="US-ASCII"?> <CreateBucketConfiguration xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><LocationConstraint>us-east-2</LocationConstraint></CreateBucketConfiguration> HTTP/1.1 200 OK x-amz-id-2: wFyFV2zNkBHpv2Zlldr+VROnnt4xTBP643l3SV7e9dq/l0oZgL97tq6PAapIO97HPC3g6zGjr80= x-amz-request-id: 9071F6FE21D5B220 Date: Tue, 11 Dec 2018 22:34:10 GMT Location: http://vadmeste-us-east-5.s3.amazonaws.com/ Content-Length: 0 Server: AmazonS3

Actually, I checked setting encoding is meaningless with S3 for incoming XML messages. So our strict nature is perhaps the problem.  Ignore the first header and unmarshal it directly is what I would do rather than bringing in character encoding libraries.

**Comment by harshavardhana** [comment]
To confirm this theory here is how one would test it

```
~ mc mb --region us-west-2 s3/vadmeste-garbage1
<?xml version="1.0" encoding="Garbage"?>
<CreateBucketConfiguration xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><LocationConstraint>us-west-2</LocationConstraint></CreateBucketConfiguration>
Bucket created successfully `s3/vadmeste-garbage1`.
```

So indeed AWS S3 simply ignores the encoding, I would suggest that we simply just do this as below on the server. Avoiding this endeavor of supporting multiple encodings. 

```diff
~ git diff
diff --git a/cmd/handler-utils.go b/cmd/handler-utils.go
index f6cf7dcf1..da6b0b710 100644
--- a/cmd/handler-utils.go
+++ b/cmd/handler-utils.go
@@ -17,6 +17,7 @@
 package cmd
 
 import (
+       "bufio"
        "bytes"
        "context"
        "io"
@@ -39,8 +40,12 @@ func parseLocationConstraint(r *http.Request) (location string, s3Error APIError
        // we do not have to validate location constraint. Bucket will
        // be created at default region.
        locationConstraint := createBucketLocationConfiguration{}
-       err := xmlDecoder(r.Body, &locationConstraint, r.ContentLength)
-       if err != nil && err != io.EOF {
+       rb := bufio.NewReader(r.Body)
+       _, _, err := rb.ReadLine()
+       if err != nil {
+               return "", ErrMalformedXML
+       }
+       if err = xmlDecoder(rb, &locationConstraint, -1); err != nil {
                logger.LogIf(context.Background(), err)
                // Treat all other failures as XML parsing errors.
                return "", ErrMalformedXML
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#dab2c0ae-117c-4697-ba72-d67b6f197172) |
<a name="dab2c0ae-117c-4697-ba72-d67b6f197172"></a>
### 6953-5e0bba2/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31367
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b0e42a0d013c:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 3 minutes and 14 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 4 seconds
(4/13) Running aws-sdk-php tests ... done in 41 seconds
(5/13) Running aws-sdk-ruby tests ... done in 6 seconds
(6/13) Running mc tests ... done in 26 seconds
(7/13) Running minio-dotnet tests ... FAILED in 2 minutes and 39 seconds

Unhandled Exception: System.AggregateException: One or more errors occurred. (Assert.Fail failed. ) ---> Microsoft.VisualStudio.TestTools.UnitTesting.AssertFailedException: Assert.Fail failed. 
   at Microsoft.VisualStudio.TestTools.UnitTesting.Assert.HandleFail(String assertionName, String message, Object[] parameters)
   at Minio.Functional.Tests.FunctionalTest.<>c.<ListIncompleteUpload_Test2>b__85_1(Exception ex) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 2138
   at System.Reactive.AutoDetachObserver`1.OnErrorCore(Exception exception) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\AutoDetachObserver.cs:line 76
   at System.Reactive.AnonymousSafeObserver`1.OnError(Exception error) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\AnonymousSafeObserver.cs:line 62
   at System.Reactive.Linq.ObservableImpl.Throw`1._.Invoke() in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\Throw.cs:line 42
   at System.Reactive.Concurrency.Scheduler.Invoke(IScheduler scheduler, Action action) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\Scheduler.Simple.cs:line 83
   at System.Reactive.Producer`2.SubscribeRaw(IObserver`1 observer, Boolean enableSafeguard) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\Producer.cs:line 127
   at System.Reactive.Linq.QueryLanguage.<>c__DisplayClass330_0`1.<Create>b__0(IObserver`1 observer) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\QueryLanguage.Creation.cs:line 46
   at System.Reactive.AnonymousObservable`1.SubscribeCore(IObserver`1 observer) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\AnonymousObservable.cs:line 37
   at System.Reactive.ObservableBase`1.ScheduledSubscribe(IScheduler _, AutoDetachObserver`1 autoDetachObserver) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\ObservableBase.cs:line 82
   at System.Reactive.Concurrency.CurrentThreadScheduler.Trampoline.Run(SchedulerQueue`1 queue) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\CurrentThreadScheduler.cs:line 126
   at System.Reactive.Concurrency.CurrentThreadScheduler.Schedule[TState](TState state, TimeSpan dueTime, Func`3 action) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\CurrentThreadScheduler.cs:line 94
   at System.Reactive.ObservableBase`1.Subscribe(IObserver`1 observer) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\ObservableBase.cs:line 51
   at Minio.Functional.Tests.FunctionalTest.ListIncompleteUpload_Test2(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 2136
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.Main(String[] args) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 241

Executed 6 out of 13 tests successfully.
```

File: cmd/utils.go [hunk 1]
```diff
@@ -105,6 +105,10 @@ const (
 	httpsScheme = "https"
 )
 
+func nopCharsetConverter(label string, input io.Reader) (io.Reader, error) {
```
Comment (harshavardhana) [inline cmd/utils.go:110]: Add a comment on why do we have this....

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/utils.go (hunk 1 above)
Comment (vadmeste) [reply cmd/utils.go:110]: > Add a comment on why do we have this....

Done

**Review by harshavardhana — DISMISSED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

File: cmd/utils.go [hunk 2]
```diff
@@ -105,6 +105,12 @@ const (
 	httpsScheme = "https"
 )
 
+// nopCharsetConvert is a dummy charset convert which just copies input to output,
```
Comment (kannappanr) [inline cmd/utils.go]: Typo `nopCharsetConvert` should be `nopCharsetConverter`

Comment (vadmeste) [reply cmd/utils.go]: > Typo `nopCharsetConvert` should be `nopCharsetConverter`

Fixed

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by kannappanr — APPROVED** [review]
LGTM. Tested, this code with an XML file with different encoding other than `utf-8`

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**